File: packages/health_connector_hk_ios/lib/src/mappers/health_record_mappers/energy_burned/active_energy_burned_record_mapper.dart
import 'package:health_connector_core/health_connector_core_internal.dart'
    show
        ActiveEnergyBurnedRecord,
        DateTimeToDto,
        Energy,
        HealthRecordId,
        sinceV1_0_0;
import 'package:health_connector_hk_ios/src/mappers/health_record_mappers/health_record_id_mapper.dart';

import 'package:health_connector_hk_ios/src/mappers/metadata_mappers/metadata_mapper.dart';
import 'package:health_connector_hk_ios/src/pigeon/health_connector_hk_ios_api.g.dart'
    show ActiveEnergyBurnedRecordDto;
import 'package:meta/meta.dart' show internal;

/// Converts [ActiveEnergyBurnedRecord] to [ActiveEnergyBurnedRecordDto].
@sinceV1_0_0
@internal
extension ActiveEnergyBurnedRecordToDto on ActiveEnergyBurnedRecord {
  ActiveEnergyBurnedRecordDto toDto() {
    return ActiveEnergyBurnedRecordDto(
      id: id.toDto(),
      startTime: startTime.millisecondsSinceEpoch,
      endTime: endTime.millisecondsSinceEpoch,
      startZoneOffsetSeconds: startTime.resolveZoneOffsetSeconds(
        startZoneOffsetSeconds,
      ),
      endZoneOffsetSeconds: endTime.resolveZoneOffsetSeconds(
        endZoneOffsetSeconds,
      ),
      metadata: metadata.toDto(),
      kilocalories: energy.inKilocalories,
    );
  }
}

/// Converts [ActiveEnergyBurnedRecordDto] to [ActiveEnergyBurnedRecord].
@sinceV1_0_0
@internal
extension ActiveEnergyBurnedRecordDtoToDomain on ActiveEnergyBurnedRecordDto {
  ActiveEnergyBurnedRecord toDomain() {
    return ActiveEnergyBurnedRecord.internal(
      id: id?.toDomain() ?? HealthRecordId.none,
      startTime: DateTime.fromMillisecondsSinceEpoch(startTime, isUtc: true),
      endTime: DateTime.fromMillisecondsSinceEpoch(endTime, isUtc: true),
      startZoneOffsetSeconds: startZoneOffsetSeconds,
      endZoneOffsetSeconds: endZoneOffsetSeconds,
      metadata: metadata.toDomain(),
      energy: Energy.kilocalories(kilocalories),
    );
  }
}

File: packages/health_connector_hc_android/android/src/main/kotlin/com/phamtunglam/health_connector_hc_android/mappers/health_record_mappers/ActiveEnergyBurnedRecordMapper.kt
package com.phamtunglam.health_connector_hc_android.mappers.health_record_mappers

import androidx.health.connect.client.records.ActiveCaloriesBurnedRecord
import androidx.health.connect.client.units.Energy
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.metadata_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.ActiveEnergyBurnedRecordDto
import java.time.Instant
import java.time.ZoneOffset

/**
 * Converts a Health Connect [ActiveCaloriesBurnedRecord] object to an [ActiveEnergyBurnedRecordDto].
 */
internal fun ActiveCaloriesBurnedRecord.toDto(): ActiveEnergyBurnedRecordDto =
    ActiveEnergyBurnedRecordDto(
        id = metadata.id,
        startTime = startTime.toEpochMilli(),
        endTime = endTime.toEpochMilli(),
        startZoneOffsetSeconds = startZoneOffset?.totalSeconds?.toLong(),
        endZoneOffsetSeconds = endZoneOffset?.totalSeconds?.toLong(),
        metadata = metadata.toDto(),
        kilocalories = energy.inKilocalories,
    )

/**
 * Converts an [ActiveEnergyBurnedRecordDto] to a Health Connect [ActiveCaloriesBurnedRecord] object.
 */
internal fun ActiveEnergyBurnedRecordDto.toHealthConnect(): ActiveCaloriesBurnedRecord =
    ActiveCaloriesBurnedRecord(
        energy = Energy.kilocalories(kilocalories),
        startTime = Instant.ofEpochMilli(startTime),
        endTime = Instant.ofEpochMilli(endTime),
        startZoneOffset = startZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        endZoneOffset = endZoneOffsetSeconds?.let { ZoneOffset.ofTotalSeconds(it.toInt()) },
        metadata = metadata.toHealthConnect(id),
    )
